grid-client/calculator/calculate.go

```diff
@@ -143,6 +152,294 @@ func calculateCU(cru, mru int64) float64 {
+		}
+		if err == ErrContractDeleted {
+			continue
+		}
+		totalCost += cost
+	}
+	return totalCost, nil
+}
+
+// Calculates the cost with a period in TFT.
+//
+// The period is the time since last updated in seconds with the provided allowance time.
+func (c *Calculator) calculatePeriodCostTFT(lastUpdatedSeconds time.Time, contract *substrate.Contract, allowance time.Duration) (float64, error) {
+	// Calculate the elapsed seconds since last billing
+	elapsedSeconds := math.Ceil(time.Since(lastUpdatedSeconds).Seconds())
+	totalPeriodSeconds := elapsedSeconds + allowance.Seconds()
+
+	contractMonthlyCostUSD, err := c.calculateContractCost(contract)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to calculate contract cost")
+	}
+
+	contractMonthlyCostTFT, err := c.USDtoTFT(contractMonthlyCostUSD)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to convert contract cost to TFT")
+	}
+	secondsPerMonth := 30 * 24 * 60 * 60 // 30 days * 24 hours * 60 minutes * 60 seconds
+	contractCostPerSecond := contractMonthlyCostTFT / float64(secondsPerMonth)
+	return contractCostPerSecond * totalPeriodSeconds, nil
+
+}
+
+// Calculates the cost of a contract per month in USD.
+func (c *Calculator) calculateContractCost(contract *substrate.Contract) (float64, error) {
+	if contract.ContractType.IsNameContract {
+		return c.calculateUniqueNameCost()
+	}
+
+	nodeID, err := getNodeID(contract)
+	if err != nil {
+		return 0, err
+	}
+
+	node, err := c.substrateConn.GetNode(nodeID)
+	if err != nil {
+		return 0, err
+	}
+
+	var nodeRentContract uint64
+
+	nodeRentContract, err = c.substrateConn.GetNodeRentContract(nodeID)
+	if err != nil && !errors.Is(err, substrate.ErrNotFound) {
+		return 0, err
+	}
+
+	if contract.ContractType.IsNodeContract {
+		return c.calculateNodeContractCost(contract, node.Certification.IsCertified, nodeRentContract > 0)
+	}
+
+	if contract.ContractType.IsRentContract {
+
+		return c.calculateRentCost(contract, *node)
+	}
+	return 0, nil
+}
+
+// Calculates the cost of a unique name per month in USD.
+func (c *Calculator) calculateUniqueNameCost() (float64, error) {
+	//TODO should we apply stacking discount?
+	pricingPolicy, err := c.substrateConn.GetPricingPolicy(defaultPricingPolicyID)
+	if err != nil {
+		return 0, err
+	}
+	// cost in unit-USD
+	monthlyCost := float64(pricingPolicy.UniqueName.Value) * 24 * 30
+	return unitToUSD(monthlyCost), nil
+}
+
+// Calculates the cost of a node contract per month in USD.
+//
+// There are two cases for node contract cost:
+//  1. Node contract on shared node: the cost of the used resources of (shared)
+//  2. Node contract on rented node: the cost of the IPV4 only if the contact includes ipv4, else it will return zero.
+func (c *Calculator) calculateNodeContractCost(contract *substrate.Contract, onCertifiedNode, isOnRentedNode bool) (float64, error) {
+	if !contract.ContractType.IsNodeContract {
+		return 0, fmt.Errorf("contract ID %d is not a node contract", contract.ContractID)
+	}
```

Comment:
can we add this check for name and rent contracts calculate functions?

Reply:
added, but this function is not being called if the contract is not a name contract
![image](https://github.com/user-attachments/assets/b1d0eb52-382e-4baa-8e7b-9bd82d4fc872)